ansi/slice.go

```diff
@@ -0,0 +1,111 @@
+package ansi
+
+import (
+	"bytes"
+
+	"github.com/charmbracelet/x/ansi/parser"
+	"github.com/rivo/uniseg"
+)
+
+// Slice slices a string to a given length, starting from X cell position.
+// This function is aware of ANSI escape codes and will not break them, and
+// accounts for wide-characters (such as East Asians and emojis).
+//
+// If a string is cut in the middle of a wide character, padding (in the
+// form of spaces) is inserted. This is done in order to maintain the width
+// of the input string.
+func Slice(s string, start int, end int) string {
+	if end < start || start == end || s == "" {
+		return ""
+	}
+
+	var cluster []byte
+	var buf bytes.Buffer
+	curPos := 0
+	pstate := parser.GroundState // initial state
+	b := []byte(s)
+	i := 0
+
+	// Here we iterate over the bytes of the string and collect printable
+	// characters and runes. We also keep track of the scan position in cells.
+	// Once we reach the given length, we start ignoring characters and only
+	// collect ANSI escape codes until we reach the end of string.
+	for i < len(b) {
```

Comment:
The only issue I see with code is the fact, this could lead to loop bomb, if a part of the code is refactored

What about adding an int like `jump` or offset, then use this

```go
i := 0
offset:= 0
for i + offset < len(b) {
       i += offset
       offset=1

// then code set the offset to something larger than 1 when needed, instead of incrementing i

}
```

Another solution might also be to simplify add something like

```go
i := 0
bomb := 0
for i < len(b) {
       bomb++
       if bomb > len(b) {
            break
       }


// keep current code in the PR unchanged

}
```

Reply:
I am not sure I understand what you mean by a "loop bomb", I have never encountered that term before. And wouldn't the unit tests catch any potential issues a refactor might bring along with it?

Reply:
An infinite loop if you prefer.

Yes, test could catch it but it can take time to understand the issue in test if they have a timeout, and could eat CPU resources on production or CI code if it gets unnoticed

Let me try to spot how it could happen

Look at like 44 or 92 the `i++`. if someone remove the line, the for loop will never end.

Another use case could be someone who adds an if just before the line 44, use a continue in it, but forget the `i++`

Reply:
It's a detail, but when I see a for without a clear way to exit the loop, I'm worried

Reply:
I understand. I will try and think of a solution

Reply:
Any solution that prevent an infinite loop is good to me. If I had provided you an idea about the offset variable it was for limiting the refactoring